### PR TITLE
Batched Newton-Raphson

### PR DESCRIPTION
**Pull request by samuelstanton**
interesting use case for Newton-Raphson is numerically solving for the optimal temperature to scale Categorical logits to a desired entropy. Could be useful to adjust sampling entropy directly instead of via temperature

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -53,10 +53,12 @@ lint.select = [
     "W",  # PYCODESTYLE WARNINGS
 ]
 
+
```
Comment (0x00b1) [inline pyproject.toml:56]: ?

File: pyproject.toml [hunk 2]
```diff
@@ -53,10 +53,12 @@ lint.select = [
...
 [tool.ruff.lint.per-file-ignores]
 "__init__.py" = [
     "F401", # MODULE IMPORTED BUT UNUSED
 ]
 
+
```
Comment (0x00b1) [inline pyproject.toml:62]: ?

File: src/beignet/_newton_raphson.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+def newton_raphson(
+    func: FuncType,
+    f_prime: FPrimeType,
+    params_batch: Tensor,
+    y_target_batch: Tensor,
+    x_init: Union[float, Tensor],
+    max_iter: int = 50,
+    tol: float = 1e-5,
+    min_bound: Optional[float] = None,
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:20]: `Number | Tensor | None`?

File: src/beignet/_newton_raphson.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    max_bound: Optional[float] = None,
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:21]: `Number | Tensor | None`?

File: src/beignet/_newton_raphson.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
+import warnings
+from typing import Any, Callable, Optional, Union
+
+import torch
+from torch import Tensor
+
+# Define function signature types
+FuncType = Callable[[Tensor, Tensor], tuple[Tensor, Any]]
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:8]: What is the `Any` in the return type?

Comment (0x00b1) [inline src/beignet/_newton_raphson.py:8]: You don’t need to add the “Type” suffix to type aliases.

Comment (0x00b1) [inline src/beignet/_newton_raphson.py:8]: This should be private (i.e., add a `_` prefix).

File: src/beignet/_newton_raphson.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+FPrimeType = Callable[[Tensor, Tensor, Any], Tensor]
+
+
+def newton_raphson(
+    func: FuncType,
+    f_prime: FPrimeType,
+    params_batch: Tensor,
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:15]: `params` is fine.

File: src/beignet/_newton_raphson.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        min_bound: Lower bound to clamp x during iteration.
+        max_bound: Upper bound to clamp x during iteration.
+        max_update_step: Maximum absolute value allowed for the x update step.
+                         Must be positive. # <-- Wrapped line 40
+
+    Returns:
+        The found roots x_batch (B,). Non-converged elements have last computed value.
+    """
+    B: int = params_batch.shape[0]
+    device: torch.device = params_batch.device
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:47]: We typically avoid explicit types.

File: src/beignet/_newton_raphson.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    dtype: torch.dtype = params_batch.dtype  # Use dtype from params for consistency
+
+    if not torch.is_tensor(y_target_batch):
+        y_target_batch = torch.tensor(y_target_batch, device=device, dtype=dtype)
+    if y_target_batch.ndim == 0:  # If single target, expand to batch
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:52]: Please add new lines between statements.

File: src/beignet/_newton_raphson.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        y_target_batch = y_target_batch.expand(B)
+    if y_target_batch.shape[0] != B:
+        raise ValueError("y_target_batch must have shape (B,) or be a scalar.")
+
+    if max_update_step <= 0:
+        raise ValueError("max_update_step must be positive.")
+
+    # Initialize x
+    x: Tensor
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:61]: ?

File: src/beignet/_newton_raphson.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        y_target_active: Tensor = y_target_batch[active]
+
+        # --- Calculate func and f_prime ---
+        try:
+            y_val: Tensor
+            aux_data: Any
+            y_prime_val: Tensor
+            y_val, aux_data = func(x_active, params_active)
+            y_prime_val = f_prime(x_active, params_active, aux_data)
+        except RuntimeError as e:
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:101]: Too vague.

File: src/beignet/_newton_raphson.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+        active_indices_update: Tensor = active_indices[update_mask]
+        if active_indices_update.numel() == 0:
+            continue
+
+        g_val_update: Tensor = g_val[update_mask]
+        y_prime_val_update: Tensor = y_prime_val[update_mask]
+
+        # Newton Step: update = error / derivative
+        raw_update: Tensor = g_val_update / (y_prime_val_update + 1e-12)
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:132]: You’re performing a massive update then clamping. It’ll, at best, converge slowly and, at worst, stall.

File: src/beignet/_newton_raphson.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    # Ensure initial x respects bounds
+    if min_bound is not None:
+        x = torch.clamp(x, min=min_bound)
+    if max_bound is not None:
+        x = torch.clamp(x, max=max_bound)
+
+    # Track active elements (those not yet converged)
+    active: Tensor = torch.ones_like(y_target_batch, dtype=torch.bool)
+
+    for i in range(max_iter):
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:81]: You need to add monotonicity and oscillation guards or it’ll ping pong between sign changes.

File: src/beignet/_newton_raphson.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        if not torch.any(active):
+            break
+
+        # --- Operate only on active elements ---
+        active_indices: Tensor = torch.where(active)[0]
+        # Ensure active_indices isn't empty before proceeding
+        if active_indices.numel() == 0:
+            break
+        x_active: Tensor = x[active]
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:90]: If `max_iter` is large (e.g., > 50), cumulative subtraction will become a problem.

File: src/beignet/_newton_raphson.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+    # Track active elements (those not yet converged)
+    active: Tensor = torch.ones_like(y_target_batch, dtype=torch.bool)
+
+    for i in range(max_iter):
+        if not torch.any(active):
+            break
+
+        # --- Operate only on active elements ---
+        active_indices: Tensor = torch.where(active)[0]
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:86]: Masks are re-computed on each step!

File: src/beignet/_newton_raphson.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+import torch
+from torch import Tensor
+
+# Define function signature types
+FuncType = Callable[[Tensor, Tensor], tuple[Tensor, Any]]
+FPrimeType = Callable[[Tensor, Tensor, Any], Tensor]
+
+
+def newton_raphson(
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:12]: A general comment, the semantics feel strange. Ideally, Newton-Raphson will respects the feasibility region or it will treat bound violations as an exception.

File: src/beignet/_newton_raphson.py (hunk 12 above)
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:81]: This is a good opportunity to experiment with `torch.func` and `vjp`.

File: src/beignet/_newton_raphson.py (hunk 11 above)
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:132]: `1e-12` will break scale-invariance for half-precision.

File: src/beignet/_newton_raphson.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+    func: FuncType,
+    f_prime: FPrimeType,
+    params_batch: Tensor,
+    y_target_batch: Tensor,
+    x_init: Union[float, Tensor],
+    max_iter: int = 50,
+    tol: float = 1e-5,
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:19]: Use both absolute and relative tolerance.

File: src/beignet/_newton_raphson.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+FPrimeType = Callable[[Tensor, Tensor, Any], Tensor]
+
+
+def newton_raphson(
+    func: FuncType,
+    f_prime: FPrimeType,
+    params_batch: Tensor,
+    y_target_batch: Tensor,
+    x_init: Union[float, Tensor],
+    max_iter: int = 50,
```
Comment (0x00b1) [inline src/beignet/_newton_raphson.py:18]: Will break if `max_iter` is `0`.

File: pyproject.toml (hunk 1 above)
Comment (samuelstanton) [reply pyproject.toml:56]: ruff format wasn't fixing my lint errors, was trying to debug (did not succeed and fixed the lines manually), will remove

File: src/beignet/_newton_raphson.py (hunk 5 above)
Comment (samuelstanton) [reply src/beignet/_newton_raphson.py:8]: additional args you may want to pass to FPrime, e.g. in the test case, class probabilities

File: src/beignet/_newton_raphson.py (hunk 11 above)
Comment (samuelstanton) [reply src/beignet/_newton_raphson.py]: def does slow convergence, was the only thing I'd found so far that prevented oscillation on my test cases.

File: src/beignet/_newton_raphson.py (hunk 12 above)
Comment (samuelstanton) [reply src/beignet/_newton_raphson.py]: yeah you can see it happen in the test case if you increase the `max_update_step` value

File: src/beignet/_newton_raphson.py (hunk 14 above)
Comment (samuelstanton) [reply src/beignet/_newton_raphson.py]: sure but `active` changes as different batch elements converge at different rates. Is this not what you'd expect?
